dev/com.ibm.ws.kernel.service/src/com/ibm/wsspi/kernel/service/utils/ConcurrentServiceReferenceSet.java

```diff
@@ -253,13 +256,56 @@ private Iterator<ConcurrentServiceReferenceElement<T>> elements() {
         return set.iterator();
     }
 
+    /**
+     * Return an iterator for the elements in reverse service ranking order.
+     */
+    private Iterator<ConcurrentServiceReferenceElement<T>> elementsAscending() {
+        Collection<ConcurrentServiceReferenceElement<T>> set;
+        synchronized (elementMap) {
+            if (elementSetUnsorted) {
+                elementSet = new ConcurrentSkipListSet<ConcurrentServiceReferenceElement<T>>(elementMap.values());
+                elementSetUnsorted = false;
+            }
+            set = elementSet;
+            // reverse the order of the elements if there are more than 1
+            if (set.size() > 1) {
+                Iterator<ConcurrentServiceReferenceElement<T>> it = set.iterator();
+                ArrayList<ConcurrentServiceReferenceElement<T>> elementSetAscending = new ArrayList<ConcurrentServiceReferenceElement<T>>();
+                while (it.hasNext()) {
+                    elementSetAscending.add(0, it.next());
+                }
+                return elementSetAscending.iterator();
+            }
+        }
+        return set.iterator();
+    }
+
+    /**
+     * Allocate and return an iterator: The iterator will
+     * return the service associated with each ServiceReference as it progresses.
+     * Creation of the iterator does not eagerly resolve services: resolution
+     * is done only once per service reference, and only when "next" would
+     * retrieve that service.
+     *
+     * @return
+     */
+    public Iterator<T> getServicesAscending() {
```

Comment:
This class is a public SPI.  So there will be other changes needed to do in order to make this change.  I think we need to do it some other way possibly.  You are saying Ascending, but in reality it is Descending.  Ascending is the default.